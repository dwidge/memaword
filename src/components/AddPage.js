import React, { useState, useRef } from "react";
import PropTypes from "prop-types";
import {
  transpose,
  dropIfIncluded,
  onlyHan,
  unique,
  uuid,
  wordsOfString,
} from "@dwidge/lib";
import { onChange } from "@dwidge/lib-react";

const AddPairs = ({ listPairs, now }) => {
  const [pairs, setpairs] = listPairs;
  const [frontWords, setfrontWords] = useState("");
  const [backWords, setbackWords] = useState("");
  const [group, setGroup] = useState("");

  const textareaFront = useRef();

  const setFrontUniqueSort = (words) =>
    setfrontWords(unique(words).sort().join("\n"));

  const onExtractWords = () => setFrontUniqueSort(wordsOfString(frontWords));

  const onExtractHan = () => setFrontUniqueSort(onlyHan([...frontWords]));

  const onCopyToClipboard = (e) => {
    textareaFront.current.select();
    document.execCommand("copy");
    // e.target.focus()
  };

  const onAddPairs = () => {
    const frontWordsA = frontWords.split("\n").filter((s) => s);
    const backWordsA = backWords.split("\n").filter((s) => s);

    if (frontWordsA.length !== backWordsA.length) {
      return;
    }

    const isSamePair = (a, b) => a.front === b.front && a.back === b.back;

    const newpairs = transpose([frontWordsA, backWordsA]).map(
      ([front, back]) => ({ id: uuid(), front, back, views: [], group })
    );

    const unseenpairs = dropIfIncluded(newpairs, pairs, isSamePair);

    setpairs(pairs.concat(unseenpairs));
    setfrontWords("");
    setbackWords("");
  };

  return (
    <div>
      <h3>Group</h3>
      <input value={group} onChange={onChange(setGroup)}></input>
      <h3>Front</h3>
      <p>Words or phrases, one per line.</p>
      <p>
        Try this: paste a large paragraph and click [Words] to extract unknown
        words from the text. [Han] will extract unknown han characters.
      </p>
      <textarea
        ref={textareaFront}
        value={frontWords}
        onChange={onChange(setfrontWords)}
        data-testid="front"
      ></textarea>
      <button onClick={onExtractWords} data-testid="buttonExtractWords">
        Words
      </button>
      <button onClick={onExtractHan} data-testid="buttonExtractHan">
        Han
      </button>
      <button onClick={onCopyToClipboard} data-testid="copyButton">
        Copy
      </button>
      <h3>Back</h3>
      <p>Words or phrases, one per line. Must match with [Front] list.</p>
      <p>
        Try this: Copy the [Front] word list to an online translator, then paste
        the translated list here.
      </p>
      <textarea
        value={backWords}
        onChange={onChange(setbackWords)}
        data-testid="back"
      ></textarea>
      <button onClick={onAddPairs} data-testid="addButton">
        Add
      </button>
    </div>
  );
};

AddPairs.propTypes = {
  listPairs: PropTypes.array.isRequired,
  now: PropTypes.number.isRequired,
};

export default AddPairs;
