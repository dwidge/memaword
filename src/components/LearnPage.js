import React, { useState } from "react";
import PropTypes from "prop-types";
import { getItemById, replaceItemById, printSecondsRounded } from "@dwidge/lib";
import { calcInterval, groupSort, reschedule } from "../lib/pairs";

const NoPairs = () => {
  return <pair-div>None to review.</pair-div>;
};

const TestPair = ({ pair, choices, onShow }) => {
  const { front } = pair;

  return (
    <pair-div>
      <pair-front data-testid="front">{front}</pair-front>
      {choices.map((v) => (
        <div key={v}>{v}</div>
      ))}
      <button data-testid="buttonShow" onClick={onShow}>
        Show
      </button>
    </pair-div>
  );
};

TestPair.propTypes = {
  pair: PropTypes.object.isRequired,
  choices: PropTypes.array.isRequired,
  onShow: PropTypes.func.isRequired,
};

const ShowPair = ({ pair, onScore }) => {
  const { front, back, views } = pair;

  const secondsPerMin = 60;
  const secondsPerDay = 24 * 60 * 60;
  const gap = calcInterval(views);

  const days = 2 * ((gap / secondsPerDay) | 0) || 1;
  const mins =
    2 *
      Math.min(secondsPerDay / secondsPerMin / 4, (gap / secondsPerMin) | 0) ||
    1;

  const scoreChoices = [
    [10, "10s", "red"],
    [secondsPerMin * mins, printSecondsRounded(secondsPerMin * mins), "orange"],
    [secondsPerDay * days, printSecondsRounded(secondsPerDay * days), "green"],
  ];

  return (
    <pair-div>
      <pair-front data-testid="front">{front}</pair-front>
      <pair-back data-testid="back">{back}</pair-back>
      <pair-choices>
        {scoreChoices.map(([time, label, color], i) => (
          <button
            key={i}
            data-testid={"buttonX" + i}
            style={{ background: color }}
            onClick={() => onScore(time)}
          >
            {label}
          </button>
        ))}
      </pair-choices>
    </pair-div>
  );
};

ShowPair.propTypes = {
  pair: PropTypes.shape({
    front: PropTypes.string,
    back: PropTypes.string,
    views: PropTypes.array,
  }).isRequired,
  onScore: PropTypes.func.isRequired,
};

const LearnPair = ({ pair, choices, onUpdate, now }) => {
  const [show, setshow] = useState(false);

  const onShow = () => {
    setshow(true);
  };

  const onScore = (t) => {
    onUpdate(reschedule(pair, t, now));
    setshow(false);
  };

  if (pair) {
    if (show) {
      return <ShowPair pair={pair} onScore={onScore} />;
    } else {
      return <TestPair pair={pair} choices={choices} onShow={onShow} />;
    }
  } else {
    return <NoPairs />;
  }
};

LearnPair.propTypes = {
  pair: PropTypes.shape({
    front: PropTypes.string,
    back: PropTypes.string,
    views: PropTypes.array,
  }),
  choices: PropTypes.array.isRequired,
  onUpdate: PropTypes.func.isRequired,
  now: PropTypes.number.isRequired,
};

const LearnPage = ({ listPairs, now }) => {
  console.log("LearnPage1", { listPairs, now });

  const [showmulti, setshowmulti] = useState(false);
  const [getlistPairs, setlistPairs] = listPairs;
  const [currentId, setcurrentId] = useState(0);
  const [choices, setchoices] = useState([]);

  const groups = groupSort(getlistPairs, now);
  const learnList = groups.review.concat(groups.new);

  if (learnList.length && !currentId) {
    setcurrentId(learnList[0].id);
    const c = showmulti
      ? learnList
          .slice(0, 5)
          .map((p) => p.back)
          .sort(() => 0.5 - Math.random())
      : [];
    setchoices(c);
  }

  const pair = getItemById(getlistPairs, currentId);

  const onUpdate = (pair) => {
    setlistPairs(replaceItemById(getlistPairs, pair));
    setcurrentId(0);
  };

  return (
    <div>
      <h3>Learn</h3>
      <p>
        {groups.review.length} review / {groups.new.length} new /{" "}
        {groups.old.length} old
      </p>
      <p>
        <input
          id="multichoice"
          type="checkbox"
          checked={showmulti}
          onChange={() => {
            setshowmulti((v) => !v);
            setcurrentId(0);
          }}
        />{" "}
        <label htmlFor="multichoice">Multichoice</label>
      </p>
      <LearnPair pair={pair} choices={choices} onUpdate={onUpdate} now={now} />
    </div>
  );
};

LearnPage.propTypes = {
  listPairs: PropTypes.array.isRequired,
  now: PropTypes.number.isRequired,
};

export default LearnPage;
