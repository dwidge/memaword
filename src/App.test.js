import React, { useState } from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import AddPage from './components/AddPage'
import LearnPage from './components/LearnPage'

const now = 60 * 1000
const listPairsInit = [
	{ id: 111, front: 'fronta', back: 'backa', views: [{ date: now, next: now + 30 * 1000 }] },
	{ id: 222, front: 'frontb', back: 'backb', views: [{ date: now, next: now + 50 * 1000 }] },
]
const getTag = (el, name) =>
	el.getElementsByTagName(name)[0]

describe('extract text', () => {
	const Test = () => {
		const listPairs = useState(listPairsInit)

		return <AddPage listPairs={listPairs} now={now + 60 * 1000} onImport={0} onExport={0} />
	}

	beforeEach(async () => {
		render(<Test />)
	})

	it('splits text into random word list', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			'Frontd? "fronte," frontf. ?',
		)
		userEvent.click(screen.getByTestId('extractButton'))
		expect(screen.getByTestId('front').value.split('\n').sort()).toEqual(['frontd', 'fronte', 'frontf'])
	})

	it('does nothing with word list', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			'Frontd? "fronte," frontf. ?',
		)
		userEvent.click(screen.getByTestId('extractButton'))
		userEvent.click(screen.getByTestId('extractButton'))
		expect(screen.getByTestId('front').value.split('\n').sort()).toEqual(['frontd', 'fronte', 'frontf'])
	})

	it.skip('extracts unique words', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			'Frontd? "fronte," frontf. Frontd frontF: frontf?',
		)
		userEvent.click(screen.getByTestId('extractButton'))
		expect(screen.getByTestId('front').value.split('\n').sort()).toEqual(['frontd', 'fronte', 'frontf'])
	})

	it.skip('extracts unknown words', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			'Fronta? "frontb," frontc.',
		)
		userEvent.click(screen.getByTestId('extractButton'))
		expect(screen.getByTestId('front').value.split('\n').sort()).toEqual(['frontc'])
	})
})

describe('submit front/back words', () => {
	const Test = () => {
		const listPairs = useState([])

		return <AddPage listPairs={listPairs} now={now + 60 * 1000} onImport={0} onExport={0} />
	}

	beforeEach(async () => {
		render(<Test />)
	})

	it('disallows mismatched words', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			['fronta', 'frontb', 'frontc'].join('\n'),
		)
		await userEvent.type(
			screen.getByTestId('back'),
			['backa', 'backb'].join('\n'),
		)
		userEvent.click(screen.getByTestId('addButton'))

		expect(screen.getByTestId('front').textContent).toEqual(['fronta', 'frontb', 'frontc'].join('\n'))
		expect(screen.getByTestId('back').textContent).toEqual(['backa', 'backb'].join('\n'))
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-front').textContent)).toEqual([])
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-back').textContent)).toEqual([])
	})

	it('matches front words to back words', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			['fronta', 'frontb', 'frontc'].join('\n'),
		)
		await userEvent.type(
			screen.getByTestId('back'),
			['backa', 'backb', 'backc'].join('\n'),
		)
		userEvent.click(screen.getByTestId('addButton'))

		expect(screen.getByTestId('front').textContent).toEqual('')
		expect(screen.getByTestId('back').textContent).toEqual('')
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-front').textContent)).toEqual(['fronta', 'frontb', 'frontc'])
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-back').textContent)).toEqual(['backa', 'backb', 'backc'])
	})

	it('ignores known pairs', async () => {
		await userEvent.type(
			screen.getByTestId('front'),
			['fronta', 'frontb', 'frontc'].join('\n'),
		)
		await userEvent.type(
			screen.getByTestId('back'),
			['backa', 'backb', 'backc'].join('\n'),
		)
		userEvent.click(screen.getByTestId('addButton'))

		await userEvent.type(
			screen.getByTestId('front'),
			['fronta', 'frontb', 'frontc'].join('\n'),
		)
		await userEvent.type(
			screen.getByTestId('back'),
			['backa', 'backb', 'backc'].join('\n'),
		)
		userEvent.click(screen.getByTestId('addButton'))

		expect(screen.getByTestId('front').textContent).toEqual('')
		expect(screen.getByTestId('back').textContent).toEqual('')
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-front').textContent)).toEqual(['fronta', 'frontb', 'frontc'])
		expect([...screen.getByTestId('pairsList').children].map(c => getTag(c, 'pair-back').textContent)).toEqual(['backa', 'backb', 'backc'])
	})
})

describe('database', () => {
	// let setlistPairs
	// setlistPairs = jest.fn().mockName('setlistPairs')
	// expect(setlistPairsSpy).toHaveBeenCalledWith(listPairsUpdated)

	const Test = () => {
		const listPairs = useState([])

		return <AddPage listPairs={listPairs} now={now + 60 * 1000} onImport={0} onExport={0} />
	}

	beforeEach(async () => {
		render(<Test />)
	})

	it.skip('imports file', () => {
		userEvent.click(screen.getByTestId('importButton'))
	})
})

describe('LearnPage', () => {
	const Test = () => {
		const listPairs = useState(listPairsInit)

		return <LearnPage listPairs={listPairs} now={now + 60 * 1000} />
	}

	beforeEach(async () => {
		render(<Test />)
	})

	it('shows front of 1st pair', () => {
		expect(screen.getByTestId('front').textContent).toEqual('fronta')
		expect(screen.queryByTestId('back')).not.toBeInTheDocument()
	})

	it('shows back of 1st pair when button clicked', () => {
		userEvent.click(screen.getByTestId('buttonShow'))
		expect(screen.getByTestId('back').textContent).toEqual('backa')
	})

	it('shows front of 2nd pair', () => {
		userEvent.click(screen.getByTestId('buttonShow'))
		userEvent.click(screen.getByTestId('buttonX1'))
		expect(screen.getByTestId('front').textContent).toEqual('frontb')
		expect(screen.queryByTestId('back')).not.toBeInTheDocument()
	})

	it('shows 3 buttons', () => {
		userEvent.click(screen.getByTestId('buttonShow'))
		expect(screen.getByTestId('buttonX0').textContent).toEqual('1s')
		expect(screen.getByTestId('buttonX1').textContent).toEqual('1m')
		expect(screen.getByTestId('buttonX2').textContent).toEqual('1d')
	})
})
