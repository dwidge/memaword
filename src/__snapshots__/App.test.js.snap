// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchPage renders 1`] = `
<div
  data-testid="pageMatch"
>
  <h3>
    Match
  </h3>
  <p>
    Choose the easiest pair.
  </p>
  <p>
    Pairs 
    <input
      max="10"
      min="1"
      step="1"
      type="range"
      value="4"
    />
     
    4
  </p>
  <pair-div>
    <pair-front
      style="display: inline-block; font-size: 2em; padding: 0.2em; margin: 0.2em;"
    >
      frontb
    </pair-front>
    <pair-front
      style="display: inline-block; font-size: 2em; padding: 0.2em; margin: 0.2em;"
    >
      fronta
    </pair-front>
    <p>
      Size 
      <input
        max="4"
        min="0.5"
        step=".5"
        type="range"
        value="2"
      />
       
      2
    </p>
    <hr />
    <pair-back
      style="display: inline-block; font-size: 2em; padding: 0.2em; margin: 0.2em;"
    >
      backb
    </pair-back>
    <pair-back
      style="display: inline-block; font-size: 2em; padding: 0.2em; margin: 0.2em;"
    >
      backa
    </pair-back>
    <p>
      Size 
      <input
        max="4"
        min="0.5"
        step=".5"
        type="range"
        value="2"
      />
       
      2
    </p>
  </pair-div>
</div>
`;
